Adds module support of includes for Rust code generation

csmulhern opened the pull request:
Adds a new `option` syntax to flatbuffers. The first usage allows setting the a `rust_module` for the schema file, to support mapping included file paths to existing modules that have been compiled for those schemas.

For more background and discussion on this approach, see: https://docs.google.com/document/d/1E1eiGlZ0DHMw5a5PvLtVffdfXHz_3Mp0_W4HAjQRPM0/edit.

This is useful for module based languages (e.g. Swift, Rust), as it allows schemas to be compiled into different units (modules) while still referencing one another.

For more background on what motivated this change, see the discussion in #8273.

Fixes #8273.

csmulhern commented:
Hey @dbaileychess @aardappel, any guidance on how I can get this reviewed?

aardappel commented:
@CasperN can you have a look?

aardappel commented:
Not being familiar with the module requirements of these languages, I don't follow why this is necessary. I'd assume `Foo` would sit in a `namespace foo_module`, and that `bar.fbs` includes `foo.fbs`, and that then this all just works as intended without command-line arguments. Where does this break, exactly?

csmulhern commented on include/flatbuffers/idl.h:
```diff
@@ -628,6 +629,25 @@ struct IncludedFile {
+// Contains a mapping from schema names to module names.
+struct ModuleMap {
+  // The type used to represent a schema name.
+  using Schema = std::string;
```
Agreed, and I am happy to remove this. I'm still not super familiar with the flatbuffers codebase, and was just emulating the style I saw here: https://github.com/google/flatbuffers/blob/cfce38ec9976ba5b53b72eeb8dcbfee9f784252e/src/idl_gen_fbs.cpp#L192.

aardappel replied:
Ah yes, I thought it must have crept in somewhere :)

csmulhern commented:
> Not being familiar with the module requirements of these languages, I don't follow why this is necessary. I'd assume `Foo` would sit in a `namespace foo_module`, and that `bar.fbs` includes `foo.fbs`, and that then this all just works as intended without command-line arguments. Where does this break, exactly?

The problem is that the module that code is a part of is an essential part of the namespace, independent of the file path, and independent of the namespace within a module. For example, with type `foo::bar::baz::Qux`...

In C++ this is all self contained:

```cpp
namespace foo {
namespace bar {
namespace baz {
  struct Qux { ... };
}  // namespace baz
}  // namespace bar
}  // namespace foo

// In any compilation unit, you can reference foo::bar::baz::Qux
```

But in Rust (and e.g. Swift), the module (or `crate` in Rust parlance) is the compilation unit, containing many sources.

```rust
# crate = foo

mod bar {
  mod baz {
    struct Qux { ... }
  }
}

// Inside this crate, you can refer to Qux as crate::bar::baz::Qux, where crate
// is a special keyword that resolves to the name of the crate this source file
// belongs to. Inside another crate, e.g. waldo, the path to refer to Qux is
// <crate_name>::path::to::Qux, or in this case, foo::bar::baz::Qux.
```

If you just typed `bar::baz::Qux`, the compiler would be looking for `baz::Qux` inside a crate named `bar`. Basically, all types are namespaced into the crate that defines that type. The problem is how to convey the top level namespace of that crate (or in more generic terms, I'm using the word module). It's not part of the flatbuffer schema. It's not part of the package declaration inside the schema. It's not part of the file path to import the schema. It's an artifact of how you choose to chunk your source files into modules.

In C++, there's no impact to splitting code into many libraries. E.g. I'm sure you're familiar with the common pattern at Google of using one `cc_library` target per source file. Every file can modify the global namespace. In Rust and Swift, unfortunately, that boundary is implicitly a namespace. So if you try to split flatbuffer source with dependenciesthat span across modules, you have a problem.

E.g. imagine the following where you take my example from the original post (a type `Bar` depends on a type `Foo` from another schema), and split it into two compilation units like so:

```bazel
fbs_library(
  name = "foo",
  srcs = ["foo.fbs"],
)

rust_fbs_library(
  name = "foo_rust",
  deps = [":foo"]
)

fbs_library(
  name = "bar",
  srcs = ["bar.fbs"],
  deps = [":foo"],
)

rust_fbs_library(
  name = "bar_rust",
  deps = [":bar"],
  rust_deps = [":bar_rust"],
)
```

The only way to refer to `Foo` from `Bar` is as `foo_rust::Foo`.

The approach taken here is to use `--module-mapping PATH=MODULE` to tell `flatc` that types inside the schema at `PATH` are inside the top level namespace `MODULE`.

To be very concrete, you can reference the test I added in the third commit.

[tests/rust_module_test/module_b.fbs](https://github.com/csmulhern/flatbuffers/blob/a2fc889a4c4a6b4f9cd7532e49574429484e5f34/tests/rust_module_test/module_b.fbs) defines a schema that depends on the type `Foo` from [tests/rust_module_test/module_a.fbs](https://github.com/csmulhern/flatbuffers/blob/a2fc889a4c4a6b4f9cd7532e49574429484e5f34/tests/rust_module_test/module_a.fbs).

The dependency between the _Rust_ code is managed through the build tool `Cargo`, using this file: https://github.com/csmulhern/flatbuffers/blob/a2fc889a4c4a6b4f9cd7532e49574429484e5f34/tests/rust_module_test/b/Cargo.toml.

Then within the Rust code for module b, we refer to `Foo` using the module name: https://github.com/csmulhern/flatbuffers/blob/a2fc889a4c4a6b4f9cd7532e49574429484e5f34/tests/rust_module_test/b/module_b_generated.rs#L56.

I really appreciate you taking the time to have a look at this. Let me know if there's any other details I can add color to.

aardappel commented:
Thanks so much for educating me :)

So I am learning that the design of FlatBuffers schemas is biased towards languages where namespaces and files are orthogonal concerns.. not surprising it started with C++.

Also, theoretically it could be made to work directly with Rust, but only if:
* Empty namespaces are forbidden.
* Revisiting/reopening namespaces in multiple schema files are forbidden.
* The code generators use the top level namespace as the crate name.
* You are supposed to output all schemas/crates in a single dir so they are parallel to eachother.

And I guess that is too much of a restriction given the amount of Rust FlatBuffers already in the wild, hence we need a workaround like this module mapping?

I guess the only question remaining is if this module mapping were to be better specified in a schema, if it can be equal for all module languages? Any reason command-line is preferrable?

```
mapping schema/foo.bs = foo_module;
```

csmulhern committed:
Adds a new option syntax for flatbuffer schemas

This syntax currently only supports setting the Rust module associated with a schema file.

csmulhern committed:
Adds rust_module to the schema definition

csmulhern committed:
Use rust_module in usage test

csmulhern committed:
Updates Rust code generator to support the rust_module option

csmulhern committed:
Regenerates Rust schema code

edvorg commented:
Hi, is there any plan to merge this ?

csmulhern commented:
@edvorg I'm happy to keep iterating on this to get it over the finish line, as needed, but we need buy-in from the maintainers. I'm reluctant to invest more without indication of a disposition to merge.
